### ISSUE TITLE
--version contains a timestamp

### ISSUE THREAD

**Issue by janhancic** [BladeRunnerJS/brjs#1520]
When I provide a `--version` argument to `serve` or `build-app` and then retrieve it with the `AppMetaService#getVersion()` I would expect to get back the value I provided. Not the value plus the timestamp. Why is that?

At the very least I would expect a method `getRawVersion()` on the service so I can have both. As right now I have to resort to parsing the version value for my actual version.

**Comment by andyberry88** [comment]
The timestamp is added to guarantee that the version is unique for every build of the app (more info on why this is useful [here](https://github.com/BladeRunnerJS/brjs/issues/1511#issuecomment-137942321)). We perhaps should change the `getVersion()` method so it returns the version up until the `-` (i.e. the `1.2.3.` part) and another that returns the whole string (`1.2.3-232334243`).

**Comment by andyberry88** [comment]
https://github.com/BladeRunnerJS/brjs/issues/1290 is also useful to explain the thought process of always adding the timestamp.

**Comment by janhancic** [comment]
Right, yeah, so we definitely need to have the ability to reliably get the supplied version. As if you're trying to display the version to the user you wouldn't want to show the timestamp as that is useless to the user.

**Comment by andyberry88** [comment]
Thinking about it I'm not sure if this is something that BRJS should be doing. You might want to only display major and minor versions to users, or only major version. Is this something BRJS should support too? IMO it's probably best for the service to provide the raw info and each app to do what it wants with the information.

**Comment by stevesouth** [comment]
Remember that our initial solution supported both options as you could use
$timestamp if you wanted a dynamic timestamp appended. I'm not sure why we
removed this feature.

On 14 September 2015 at 12:41, Andy Berry <notifications@github.com> wrote:

> Thinking about it I'm not sure if this is something that BRJS should be
> doing. You might want to only display major and minor versions to users, or
> only major version. Is this something BRJS should support too? IMO it's
> probably best for the service to provide the raw info and each app to do
> what it wants with the information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BladeRunnerJS/brjs/issues/1520#issuecomment-140046497>
> .
>



-- 
*Stephen South*
Professional Services Development Manager | Caplin Systems Ltd
Office: +44 (0)20 7826 9618
Mobile: +44 (0)7833 240 370

www.caplin.com
<https://mail.caplin.com/owa/redir.aspx?C=ycOKrX3n8UOuU1SUZqv1hZTBVvrohNEI3NDsCJnyTotUoUMYoptmHmtnEajb7p8dfsEKr963Mak.&URL=http%3a%2f%2fwww.caplin.com%2f>

Single-Dealer Platform News Feed: *SerenDiPity*
<https://mail.caplin.com/owa/redir.aspx?C=ycOKrX3n8UOuU1SUZqv1hZTBVvrohNEI3NDsCJnyTotUoUMYoptmHmtnEajb7p8dfsEKr963Mak.&URL=http%3a%2f%2fsingledealerplatforms.wordpress.com%2f>
Caplin Tech Blog: *Platformability
<https://mail.caplin.com/owa/redir.aspx?C=ycOKrX3n8UOuU1SUZqv1hZTBVvrohNEI3NDsCJnyTotUoUMYoptmHmtnEajb7p8dfsEKr963Mak.&URL=http%3a%2f%2fblog.caplin.com%2f>*

**Comment by andyberry88** [comment]
The app cache plugin now uses the same version mechanism as the rest of BRJS, set via the single `--version` argument. For various reasons we've chosen to automatically append the timestamp to ensure the version is unique. The app meta service provides the raw version data. Surely it's trivial to write a service that provides access to the part of the version string you are interested in? Each app's version may be different (it could be separated by `-` rather than `.` for example), IMO it makes sense for that to be an app concern not a framework concern.

**Comment by janhancic** [comment]
Umm, but that's the whole point. BRJS changes the thing I pass in. You can't change my value and then call that the raw value.

I'm not asking for BRJS to support getting minor versions or anything like that. I just want an easy way to get my actual version value, without whatever BRJS appends. That's it's concern and I shouldn't even be aware of it to be honest.

So, if I run:

```
brjs serve --version my_Version.Number-foo
```

Then `AppMetaService` should return me `my_Version.Number-foo`, not `my_Version.Number-foo-TIMESTAMP`. I honestly don't know why the latter would be ever useful to me on the frontend.

**Comment by dchambers** [comment]
> I'm not asking for BRJS to support getting minor versions or anything like that. I just want an easy way to get my actual version value, without whatever BRJS appends. That's it's concern and I shouldn't even be aware of it to be honest.

@janhancic, this sounds reasonable to me.

**Comment by andyberry88** [comment]
Ok. I'm not sure 'raw' is the right word. How about changing `getVersion()` to return the version *without* the timestamp, unless there is no user entered version in which case it returns the timestamp, and a `getFullVersion()` which returns the entire version string?

**Comment by janhancic** [comment]
I would prefer this behaviour:

- `getVersion()` returns either what I passed to `--version` or `dev` if nothing was passed
- `getFullVersion()` returns the entered version (or dev) plus `-TIMESTAMP`

**Comment by sospirited** [comment]
This feature from a mechanism point of view could be made generic so that it's possible to pass in a map of key/value items when `build-app` is called.

Specifically, for additional meta information specific to that snapshot version of the application.

This can be particularly useful for development teams to easily make available key/values when doing continuous deployment to show in-app.

Things like:
- version
- timestamp (implicit?)
- commit hash
- sprint
etc

As an example, a CI build could be configured to run
E.g. `brjs build-app <myapp> -w --meta "version:1.0.0, commit: abcdefg, sprint: 32"

And have client code which could then decide to display what it wants (even conditionally based on JNDI tokens for whether it's in dev/prod environments etc).

Not saying the above is perfect, but that the --version flag could call this more generic args option under the hood.

The use case we have at the moment is we'd like to be able to know what commit hash a deployment was built from, in our app which we have to to hacky code to abuse the --version feature which may change as the result of this convo.
